api/turing/api/router_versions_api.go

```diff
@@ -24,11 +24,31 @@ func (c RouterVersionsController) ListRouterVersions(
 		return errResp
 	}
 
+	var project *mlp.Project
+	if project, errResp = c.getProjectFromRequestVars(vars); errResp != nil {
+		return errResp
+	}
+
 	// List router versions
 	routerVersions, err := c.RouterVersionsService.ListRouterVersions(router.ID)
 	if err != nil {
 		return InternalServerError("unable to retrieve router versions", err.Error())
 	}
+
+	// Append monitoring URL to router versions
+	for _, routerVersion := range routerVersions {
+		monitoringURL, err := c.RouterVersionsService.GenerateMonitoringURL(
```

Comment:
Is there any particular reason to have this done at the controller level, instead of a service level (similarly how it's done for ensembling jobs)?

Reply:
Also, kinda nitpicking, but you can avoid creating a new `monitoringURL` variable for each item in the slice:
```golang
var err error
for _, routerVersion := range routerVersions {
  routerVersion.MonitoringURL, err = c.RouterVersionsService.GenerateMonitoringURL(...)
  if err != nil {
    return InternalServerError("unable to generate monitoringURL for router version", err.Error()). 
  }
}
```

Reply:
Yeah, actually I feel that it's kinda weird to just abstract the for loop out into a service. It looks long to you because of the linter (I have to split the 1 function call into 6 lines). Your example below shows how simple the line is already, it's already in a service. If I were to do it in the service, all it does is a for loop, add an additional dependency, router versions service into the router service and process this again. For now the services looks simple but when things get more complicated, we may run into a dependency cycle in the future. i.e. Service A depends on Service B and Service B depends on Service C. What if Service C requires a dependency on service A or it's just mentally difficult to process all the dependencies.

I would rather push for a mediator pattern that is decoupled from the database calls than to push this to the service.

Reply:
But I also understand why you would want it there:
```go
	if router, errResp = c.getRouterFromRequestVars(vars); errResp != nil {
		return errResp
	}
	if routerVersion, errResp = c.getRouterVersionFromRequestVars(vars); errResp != nil {
		return errResp
	}
```

This should have handled everything. Anyway, I'll do it but this might be a problem next time.

Reply:
Done.